src/cmd/tools/doctest/doctest.go

```diff
@@ -0,0 +1,171 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+const (
+	operationRegEx  = `(?s)<operation>.*?</operation>|(?s)<operation_json>.*?</operation_json>`
+	validationRegEx = `(?s)<validation>.*?</validation>`
+
+	operationOpen      = `<operation>`
+	operationClose     = `</operation>`
+	operationJSONOpen  = `<operation_json>`
+	operationJSONClose = `</operation_json>`
+	validationOpen     = `<validation>`
+	validationClose    = `</validation>`
+
+	all = `<operation>|</operation>|<validation>|</validation>`
+)
+
+var (
+	file string
+)
+
+func init() {
+	flag.StringVar(&file, "file", "", "File to be parsed")
+}
+
+func main() {
+	flag.Parse()
+
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Fatalf("unable to read source file: %v\n", err)
+	}
+
+	validationRegEx, operationRegEx, err := bashScript(f, file)
+	if err != nil {
+		log.Fatalf("unable to create bash script: %v", err)
+	}
+
+	if err := markdownFile(f, file, validationRegEx, operationRegEx); err != nil {
+		log.Fatalf("unable to create markdown file: %v", err)
+	}
+}
+
+func markdownFile(contents []byte, fileName string, validationRegEx *regexp.Regexp, opRegex *regexp.Regexp) error {
+	// create new markdown file based on name of source file
+	out, err := os.Create(strings.Replace(fileName, ".source", "", -1))
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	out.Chmod(0644)
+
+	// find all of the operations in the source file
+	matches := opRegex.FindAll(contents, 100)
+	for _, match := range matches {
+		validations := validationRegEx.FindAll(match, 100)
+		if len(validations) == 0 {
+			continue
+		}
+
+		// remove the validations and if the operation block is empty, replace it with nothing
+		noValidationBlock := bytes.Replace(match, validations[0], nil, -1)
+
+		if string(noValidationBlock) == fmt.Sprintf(operationOpen+"\n"+operationClose) {
+			contents = bytes.Replace(contents, match, []byte(""), -1)
+		}
+	}
+
+	// convert tags to markdown syntax
+	OpOpen := regexp.MustCompile(operationOpen)
+	OpClose := regexp.MustCompile(operationClose)
+	OpJSONOpen := regexp.MustCompile(operationJSONOpen)
+	OpJSONClose := regexp.MustCompile(operationJSONClose)
+
+	contents = OpOpen.ReplaceAll(contents, []byte("```"))
+	contents = OpClose.ReplaceAll(contents, []byte("```"))
+	contents = OpJSONOpen.ReplaceAll(contents, []byte("```json"))
+	contents = OpJSONClose.ReplaceAll(contents, []byte("```"))
+	contents = validationRegEx.ReplaceAll(contents, nil)
+	out.Write(contents)
+
+	return nil
+}
+
+func bashScript(contents []byte, fileName string) (*regexp.Regexp, *regexp.Regexp, error) {
+	// create bash script file based on name of source file
+	scriptFileName := fmt.Sprintf("%s%s", strings.Replace(fileName, ".md.source", "", -1), ".sh")
+	script, err := os.Create(scriptFileName)
```

Comment:
use https://golang.org/pkg/os/#OpenFile (look at their examples) to create a file with the right permissions